Update readme (#1619)
gordon-cs/gordon-360-ui: docs/Developer's Guide.md

# Developer's Guide

This guide walks you through the basics of developing the Gordon 360 frontend, all the way from setting up an initial development environment to deploying changes to production.

# Contents

- [Getting Started](#getting-started)
- [Connecting to the Backend](#connecting-to-the-backend)
  - [Connecting to Remote Backend via SSH](#connecting-to-remote-backend-via-ssh)
- [Code Style](#code-style)
- [Project File Organization](#file-organization)
  - [Components](#components)
  - [Contexts and Hooks](#contexts-and-hooks)
  - [Services](#services)
  - [Views](#views)
- [Libraries](#libraries)
- [Continuous Integration and Deployment](#continuous-integration-and-deployment)

# Getting Started

> Here are the steps to setupthe frontend for development:

Reviewer: ```suggestion
Here are the steps to setup the frontend for development:
```

1. Clone this repository to the machine that you will develop on.

1. Ensure that [NodeJS](https://nodejs.org/en/) is installed on that machine.

   From a terminal, run this command:

   ```
   node -v
   ```

   If it returns a version number, then NodeJS is installed. Proceed to the next step.

   If it doesn't return a number, you will need to install NodeJS. There are two ways to do this:

   1. Download and run the installer from [the NodeJS Website](https://nodejs.org/en/download/)
   1. Use a tool that manages Node installations, such as
      - For any OS: [Fast Node Manager](https://github.com/Schniz/fnm)
      - For MacOS/Linux: [Node Version Manager](https://github.com/nvm-sh/nvm)
      - For Windows: [NVM for Windows](https://github.com/coreybutler/nvm-windows)

1. Install dependencies

   From a terminal, go to the directory where you cloned the project (e.g. the `gordon-360-ui` folder) and run the following command:

   ```
   npm install
   ```

   This will install the frontend's dependencies. It may take a long time the first time you run it.

1. Start the project

   From a terminal in the project directory, run this command:

   ```

>    npm run start

Reviewer: ```suggestion
   npm start
```
Can't this be simplified to just the above?

Author: Sure, but they need to do `npm run ____` for all the other commands in `package.json`, so I figured for consistency's sake I'd leave it as `npm run start`.

### Steps:

1. Your local machine must be configured as an SSH host

   [Windows Installation](https://learn.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse?tabs=gui).

   [Ubuntu Installation](ubuntu.com/server/docs/service-openssh)

   MacOS & other Linux distributions should already have an SSH server installed.

2. Create the SSH tunnel

   After setting up the ssh server on your host machine, find your host IP or DNS address. Open a command prompt or terminal window and enter the following:

   ```
   hostname
   ```

   Log into the VM and open a command prompt/powershell window. n command prompt/terminal window enter the following command,

   ```
   ssh -R localhost:[API_VM_PORT_NUMBER]:localhost:[API_PC_PORT_NUMBER] [USERNAME]@[IP or HOSTNAME]
   ```

   where:

   - `API_VM_PORT_NUMBER` is the port on which the 360 API is running on the VM.
   - `API_PC_PORT_NUMBER` is the port that you want the API to be sent to on your personal computer (not the VM).
   - `USER` is your account on the host machine.
   - `IP or HOSTNAME` is the IP or HOSTNAME of your host machine.

> You are now able to point your local frontend to the remote backend by setting `REACT_APP_API_URL` to `localhost:[API_PC_PORT_NUMBER]`.

Reviewer: Did you intend to have this in a separate file? The link on the first line points to the SSH backend file, but the details are all here anyway.

Author: Ah, good catch. I merged them both into the same file but forgot to remove the link.

# File Organization

The source files for the app are in `./src`. The other top-level folders are as follows:

- `.github` contains configuration for GitHub Actions, which is used for maintenance, security, and CI/CD tasks.
- `.husky` contains pre-commit hooks to enforce consistent formatting.
- `.vscode` contains configuration for Visual Studio Code
- `docs` contains supplemental documentation.
- `build` contains the built application; not tracked by Git
- `node_modules` contains dependencies installed by `npm`; not tracked by Git
- `public` contains assets that should not be processed by Webpack. Only files inside `public` can be used from `public/index.html`.

The structure of the `src` directory is as follows:

```plain
├── components
│   └── ...
├── contexts
│   └── ...
├── hooks
│   └── ...
├── services
│   └── ...
├── views
│   └── ...
├── app.js
├── index.js
└── pwa.js
```

## Components

```plain
components
├── ...
├── EventList
│   ├── components
│   │   └── ...
│   ├── EventList.module.scss
│   └── index.js
└── ...
```

> This folder contains components that, when used together, make up the views of the application. Each component should be small and focus on doing One Thing Well. Read about the [Atomic Design Methodology](http://atomicdesign.bradfrost.com/chapter-2/) for a useful perspective on writing small, reusable components and composing them into larger areas of functionality.

Reviewer: ```suggestion
This folder contains components that, when used together, make up the views of the application. Each component should be small and focus on doing **one thing well**. Read about the [Atomic Design Methodology](http://atomicdesign.bradfrost.com/chapter-2/) for a useful perspective on writing small, reusable components and composing them into larger areas of functionality.
```

> Each component must have a folder named in PascalCase (also known as upper camel case) containing a file called `index.js`. Using that filename allows the component to be imported by folder, instead of by file: `import MyComponent from 'components/MyComponent` instead of `import MyComponent from 'components/MyComponent/my-component`.

Reviewer: ```suggestion
Each component must have a folder named in PascalCase format (also known as upper camel case) containing a file called `index.js`. Using that filename allows the component to be imported by folder, instead of by file: `import MyComponent from 'components/MyComponent` instead of `import MyComponent from 'components/MyComponent/my-component`.
```

Reviewer: Okay, sure. The real issue in my mind is that "...folder named in PascalCase..." seems awkward to me. If you don't think so, fine to leave it, but perhaps 
```suggestion
Each component must have a folder with its name in PascalCase (also known as upper camel case) containing a file called `index.js`. Using that filename allows the component to be imported by folder, instead of by file: `import MyComponent from 'components/MyComponent` instead of `import MyComponent from 'components/MyComponent/my-component`.
```
